src/tool/subcommands/api_cmd.rs

```diff
@@ -1228,4 +1241,29 @@ mod tests {
             .reject("Chain".to_string());
         assert!(!list.authorize("Filecoin.ChainGetBlock"));
     }
+
+    #[test]
+    fn test_derive_protocol() {
+        let forest = ApiInfo::from_str("/ip4/127.0.0.1/tcp/2345/http").expect("infallible");
+        let lotus = ApiInfo::from_str("/ip4/127.0.0.1/tcp/1234/http").expect("infallible");
+        assert!(matches!(
+            derive_protocol(&forest, &lotus),
+            Ok(CommunicationProtocol::Http)
+        ));
+
+        let forest = ApiInfo::from_str("/ip4/127.0.0.1/tcp/2345/ws").expect("infallible");
+        let lotus = ApiInfo::from_str("/ip4/127.0.0.1/tcp/1234/ws").expect("infallible");
+        assert!(matches!(
+            derive_protocol(&forest, &lotus),
+            Ok(CommunicationProtocol::Ws)
+        ));
+
+        let forest = ApiInfo::from_str("/ip4/127.0.0.1/tcp/2345/http").expect("infallible");
+        let lotus = ApiInfo::from_str("/ip4/127.0.0.1/tcp/1234/ws").expect("infallible");
+        assert!(derive_protocol(&forest, &lotus).is_err());
+
+        let forest = ApiInfo::from_str("/ip4/127.0.0.1/tcp/2345/wss").expect("infallible");
+        let lotus = ApiInfo::from_str("/ip4/127.0.0.1/tcp/1234/wss").expect("infallible");
+        assert!(derive_protocol(&forest, &lotus).is_err());
```

Comment:
nit: I'd unwrap on those, `infallible` doesn't bring any more context, and these are tests anyway